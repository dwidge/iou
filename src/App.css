* {
box-sizing: border-box;
}

nav {
display: flex;
justify-content: space-between;
overflow: scroll;
}

.page{
overflow: scroll;
}

.link,.link-active {
color: inherit;
text-decoration: inherit;
border: solid 1px orange;
padding: 1em;
display: block;
}

.link-active {
background: orange;
}

button {
padding: 1em;
margin: 0em;
border: solid 1px orange;
}

textarea {
width: 100%;
height: 10em;
}
