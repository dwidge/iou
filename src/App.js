import React, { useState, useEffect } from 'react'
import { Storage } from '@dwidge/lib-react'

import Clients from './components/Clients'
import Invoices from './components/Invoices'
import Receipts from './components/Receipts'
import Bank from './components/Bank'
import Statements from './components/Statements'
import Settings from './components/Settings'
import Navbar from './components/Navbar'
import { newRef, useKeyVal } from './components/lib'

// import './App.css'
const { useStorage } = Storage(useState, useEffect)

const defaultSettings = {
	id: 1,
	clnpre: 'C',
	clnnext: 1,
	stmpre: 'S',
	stmnext: 1,
	recpre: 'R',
	recnext: 1,
	invpre: 'I',
	invnext: 1,
}

const App = () => {
	const stClients = useStorage('iou_clients', [])
	const stInvoices = useStorage('iou_invoices', [])
	const stReceipts = useStorage('iou_receipts', [])
	const stSettings = useKeyVal(useStorage('iou_settings', {}), defaultSettings)

	return (
		<Navbar brand='IoU' pages={{
			'/clients': { name: 'Clients', element: (<Clients stClients={stClients} newId={newRef(stSettings)('clnpre', 'clnnext')} />) },
			'/invoices': { name: 'Invoices', element: (<Invoices stInvoices={stInvoices} aClients={stClients[0]} stSettings={stSettings} />) },
			'/receipts': { name: 'Receipts', element: (<Receipts stReceipts={stReceipts} aClients={stClients[0]} stInvoices={stInvoices} stSettings={stSettings} />) },
			'/bank': { name: 'Bank', element: (<Bank stClients={stClients} stInvoices={stInvoices} stSettings={stSettings} stReceipts={stReceipts} newClientId={newRef(stSettings)('clnpre', 'clnnext')}/>) },
			'/statements': { name: 'Statements', element: (<Statements aClients={stClients[0]} aInvoices={stInvoices[0]} aReceipts={stReceipts[0]} stSettings={stSettings} />) },
			'/settings': { name: 'Settings', element: (<Settings stSettings={stSettings} />) },
		}}/>
	)
}

export default App
