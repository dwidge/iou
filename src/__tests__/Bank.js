import React from 'react'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import {
	MemoryRouter,
} from 'react-router-dom'
// import { debug } from 'jest-preview';

import * as J from '@dwidge/lib-react'

import App from '../App'

const clear = J.clear(userEvent, screen)
const type = J.type(userEvent, screen)
const click = J.click(userEvent, screen)
const input = async (id, text) => {
	await clear(id)
	await type(id, text)
}
const snap = id =>
	expect(screen.getByTestId(id)).toMatchSnapshot()

jest.mock('@dwidge/lib', () => ({
	...jest.requireActual('@dwidge/lib'),
	uuid: ((value = 10) => () => String(++value))(),
}))
afterEach(() => {
	jest.restoreAllMocks()
})

describe('Bank', () => {
	test('renders page', () => {
		render(<MemoryRouter initialEntries={['/statements']}>
			<App />
		</MemoryRouter>)

		snap('tableStatements')
	})

	test('detects patterns', async () => {
		render(<MemoryRouter initialEntries={['/clients']}>
			<App />
		</MemoryRouter>)

		await click('buttonAdd')
		await input('inputName', 'clientA')
		await input('inputPatterns', 'dd')
		await click('buttonSave')
		await click('buttonAdd')
		await input('inputName', 'clientB')
		await click('buttonSave')
		await click('nav/bank')
		await input('inputBank statement', `
22/33/4444 dd 8.8 * 1,109,999.99
22/33/4444 dd 7.7 * 999.99
11/22/3333 aa 4 bb 5.5 cc 6.66 7.77
`)
		snap('route/bank')
		await click('buttonEdit146')
		await input('inputViewClient', 'clientB')
		await click('buttonSave')
		await click('buttonAddAll')
		snap('route/bank')
		await click('nav/receipts')
		snap('route/receipts')
		await click('nav/clients')
		snap('route/clients')
	})
})
