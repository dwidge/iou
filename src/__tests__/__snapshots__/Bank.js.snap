// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bank detects patterns 1`] = `
<div
  data-testid="route/bank"
>
  <p>
    Access a bank statement, press Ctrl+A, press Ctrl+C, then return here and click in the box below, press Ctrl+V.
  </p>
  <form
    class=""
  >
    <div
      class="mb-3"
    >
      <label
        class="form-label"
      >
        Bank statement
      </label>
      <textarea
        class="form-control"
        data-testid="inputBank statement"
        rows="5"
      >
        
22/33/4444 dd 8.8 * 1,109,999.99
22/33/4444 dd 7.7 * 999.99
11/22/3333 aa 4 bb 5.5 cc 6.66 7.77

      </textarea>
    </div>
  </form>
  <div-page
    data-testid="tableStatements"
  >
    <div
      class="table-responsive"
    >
      <table
        class="table table-striped table-bordered table-hover"
      >
        <thead>
          <tr>
            <th>
              Client
            </th>
            <th>
              Date
            </th>
            <th>
              Desc
            </th>
            <th>
              In/Out
            </th>
            <th>
              Balance
            </th>
            <th>
              Match
            </th>
            <th>
              Pattern
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <column-text>
                clientA
              </column-text>
            </td>
            <td>
              <column-text>
                22/33/4444
              </column-text>
            </td>
            <td>
              <column-text>
                dd
              </column-text>
            </td>
            <td>
              <column-text>
                8.80
              </column-text>
            </td>
            <td>
              <column-text>
                1109999.99
              </column-text>
            </td>
            <td>
              <column-text>
                C001
              </column-text>
            </td>
            <td>
              <column-text>
                dd
              </column-text>
            </td>
            <td>
              <button
                class="btn btn-primary"
                data-testid="buttonEdit146"
                type="button"
              >
                Edit
              </button>
              <button
                class="btn btn-primary"
                data-testid="buttonDel146"
                type="button"
              >
                Del
              </button>
            </td>
          </tr>
          <tr>
            <td>
              <column-text>
                clientA
              </column-text>
            </td>
            <td>
              <column-text>
                22/33/4444
              </column-text>
            </td>
            <td>
              <column-text>
                dd
              </column-text>
            </td>
            <td>
              <column-text>
                7.70
              </column-text>
            </td>
            <td>
              <column-text>
                999.99
              </column-text>
            </td>
            <td>
              <column-text>
                C001
              </column-text>
            </td>
            <td>
              <column-text>
                dd
              </column-text>
            </td>
            <td>
              <button
                class="btn btn-primary"
                data-testid="buttonEdit147"
                type="button"
              >
                Edit
              </button>
              <button
                class="btn btn-primary"
                data-testid="buttonDel147"
                type="button"
              >
                Del
              </button>
            </td>
          </tr>
          <tr>
            <td>
              <column-text>
                -
              </column-text>
            </td>
            <td>
              <column-text>
                11/22/3333
              </column-text>
            </td>
            <td>
              <column-text>
                aa 4 bb 5.5 cc
              </column-text>
            </td>
            <td>
              <column-text>
                6.66
              </column-text>
            </td>
            <td>
              <column-text>
                7.77
              </column-text>
            </td>
            <td>
              <column-text />
            </td>
            <td>
              <column-text>
                aa * bb *.* cc
              </column-text>
            </td>
            <td>
              <button
                class="btn btn-primary"
                data-testid="buttonEdit148"
                type="button"
              >
                Edit
              </button>
              <button
                class="btn btn-primary"
                data-testid="buttonDel148"
                type="button"
              >
                Del
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <button
      class="btn btn-primary"
      data-testid="buttonAdd"
      type="button"
    >
      Add
    </button>
    <button
      class="btn btn-primary"
      data-testid="buttonClear"
      type="button"
    >
      Clear
    </button>
  </div-page>
  <button
    class="btn btn-primary"
    data-testid="buttonAddAll"
    type="button"
  >
    Add All to Receipts
  </button>
  <p
    class="mt-3"
  >
    Any rows with blank Client will be deleted. Discarded rows are added to ignore patterns.
  </p>
  <p>
    Any rows already in Receipts (with same Client, Date and In/Out) are duplicates. They will not be added again.
  </p>
  <form
    class=""
  >
    <div
      class="mb-3"
    >
      <label
        class="form-label"
      >
        Ignore patterns
      </label>
      <textarea
        class="form-control"
        data-testid="inputIgnore patterns"
        rows="5"
      />
    </div>
  </form>
</div>
`;

exports[`Bank detects patterns 2`] = `
<div
  data-testid="route/bank"
>
  <div
    class="fade alert alert-success show"
    role="alert"
  >
    2 added. 0 duplicated. 1 discarded.
  </div>
  <p>
    Access a bank statement, press Ctrl+A, press Ctrl+C, then return here and click in the box below, press Ctrl+V.
  </p>
  <form
    class=""
  >
    <div
      class="mb-3"
    >
      <label
        class="form-label"
      >
        Bank statement
      </label>
      <textarea
        class="form-control"
        data-testid="inputBank statement"
        rows="5"
      />
    </div>
  </form>
  <div-page
    data-testid="tableStatements"
  >
    <div
      class="table-responsive"
    >
      <table
        class="table table-striped table-bordered table-hover"
      >
        <thead>
          <tr>
            <th>
              Client
            </th>
            <th>
              Date
            </th>
            <th>
              Desc
            </th>
            <th>
              In/Out
            </th>
            <th>
              Balance
            </th>
            <th>
              Match
            </th>
            <th>
              Pattern
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
    <button
      class="btn btn-primary"
      data-testid="buttonAdd"
      type="button"
    >
      Add
    </button>
    <button
      class="btn btn-primary"
      data-testid="buttonClear"
      type="button"
    >
      Clear
    </button>
  </div-page>
  <button
    class="btn btn-primary"
    data-testid="buttonAddAll"
    type="button"
  >
    Add All to Receipts
  </button>
  <p
    class="mt-3"
  >
    Any rows with blank Client will be deleted. Discarded rows are added to ignore patterns.
  </p>
  <p>
    Any rows already in Receipts (with same Client, Date and In/Out) are duplicates. They will not be added again.
  </p>
  <form
    class=""
  >
    <div
      class="mb-3"
    >
      <label
        class="form-label"
      >
        Ignore patterns
      </label>
      <textarea
        class="form-control"
        data-testid="inputIgnore patterns"
        rows="5"
      >
        aa * bb *.* cc
      </textarea>
    </div>
  </form>
</div>
`;

exports[`Bank detects patterns 3`] = `
<div
  data-testid="route/receipts"
>
  <div-page
    data-testid="tableReceipts"
  >
    <div
      class="table-responsive"
    >
      <table
        class="table table-striped table-bordered table-hover"
      >
        <thead>
          <tr>
            <th>
              Ref
            </th>
            <th>
              Date
            </th>
            <th>
              Client
            </th>
            <th>
              Total
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <column-text>
                R001
              </column-text>
            </td>
            <td>
              <column-text>
                22/33/4444
              </column-text>
            </td>
            <td>
              <column-text>
                clientB
              </column-text>
            </td>
            <td>
              <column-text>
                8.80
              </column-text>
            </td>
            <td>
              <button
                class="btn btn-primary"
                data-testid="buttonEdit149"
                type="button"
              >
                Edit
              </button>
              <button
                class="btn btn-primary"
                data-testid="buttonDel149"
                type="button"
              >
                Del
              </button>
            </td>
          </tr>
          <tr>
            <td>
              <column-text>
                R001
              </column-text>
            </td>
            <td>
              <column-text>
                22/33/4444
              </column-text>
            </td>
            <td>
              <column-text>
                clientA
              </column-text>
            </td>
            <td>
              <column-text>
                7.70
              </column-text>
            </td>
            <td>
              <button
                class="btn btn-primary"
                data-testid="buttonEdit150"
                type="button"
              >
                Edit
              </button>
              <button
                class="btn btn-primary"
                data-testid="buttonDel150"
                type="button"
              >
                Del
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <button
      class="btn btn-primary"
      data-testid="buttonAdd"
      type="button"
    >
      Add
    </button>
    <button
      class="btn btn-primary"
      data-testid="buttonClear"
      type="button"
    >
      Clear
    </button>
  </div-page>
</div>
`;

exports[`Bank detects patterns 4`] = `
<div
  data-testid="route/clients"
>
  <div-page
    data-testid="tableClients"
  >
    <div
      class="table-responsive"
    >
      <table
        class="table table-striped table-bordered table-hover"
      >
        <thead>
          <tr>
            <th>
              Ref
            </th>
            <th>
              Name
            </th>
            <th>
              Phone
            </th>
            <th>
              Patterns
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <column-text>
                C001
              </column-text>
            </td>
            <td>
              <column-text>
                clientA
              </column-text>
            </td>
            <td>
              <column-text>
                555
              </column-text>
            </td>
            <td>
              <column-text>
                dd
              </column-text>
            </td>
            <td>
              <button
                class="btn btn-primary"
                data-testid="buttonEdit11"
                type="button"
              >
                Edit
              </button>
              <button
                class="btn btn-primary"
                data-testid="buttonDel11"
                type="button"
              >
                Del
              </button>
            </td>
          </tr>
          <tr>
            <td>
              <column-text>
                C002
              </column-text>
            </td>
            <td>
              <column-text>
                clientB
              </column-text>
            </td>
            <td>
              <column-text>
                555
              </column-text>
            </td>
            <td>
              <column-text>
                
dd
              </column-text>
            </td>
            <td>
              <button
                class="btn btn-primary"
                data-testid="buttonEdit12"
                type="button"
              >
                Edit
              </button>
              <button
                class="btn btn-primary"
                data-testid="buttonDel12"
                type="button"
              >
                Del
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <button
      class="btn btn-primary"
      data-testid="buttonAdd"
      type="button"
    >
      Add
    </button>
    <button
      class="btn btn-primary"
      data-testid="buttonClear"
      type="button"
    >
      Clear
    </button>
  </div-page>
</div>
`;

exports[`Bank renders page 1`] = `
<div-page
  data-testid="tableStatements"
>
  <div
    class="table-responsive"
  >
    <table
      class="table table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th>
            ClientRef
          </th>
          <th>
            BalanceDue
          </th>
          <th>
            Send
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</div-page>
`;
